### DISCUSSION TITLE
Protocol versioning

### DISCUSSION THREAD

**Discussion by kulakowski-wasm** [bytecodealliance/SIG-Registries#14] [General]
Several discussions have alluded to protocol versioning as a way to adapt APIs we publish over time. For instance, we may define a new version of the Publish API in the future.

I think this topic is worth discussing now for a few reasons. I think protocol versioning design decisions may affect:

1. protocols per se

This might be as straightforward as subsuming ad-hoc versioning that would otherwise be in an API. It might be a design escape hatch for us to get an MVP out without accounting for a future need. Or something more subtle!

2. relationships between protocols

To this point, we've talked about things as separate artifacts (to keep picking on this one: "Publish API"). Some of this is a natural breaking up of the problem space into smaller parts. Some is a reflection of the fact that different entities will use different parts, or different implementations of a part. I think breaking this down in terms of what we think we will *actually ship* as separable entities will be time well spent!

Consider two widely different versioning strategies: independently version every protocol; globally version all protocols together.

Which is a pro or a con depending on your philosophy on updating software I suppose :)

3. lifetime of protocol versions

I sure do have a feeling that we will ship a thing, and then wish that people would stop using the first version thing.

Exactly at what granularity and boundaries we will hope to do so is part of this design discussion, I think.

**Comment by lann** [comment]
> I think breaking this down in terms of what we think we will actually ship as separable entities will be time well spent!

Agreed; I have been mentally organizing APIs into 3 groups along system architecture lines:

1. Maintainer APIs: publish release, yank release
2. Consumer APIs: get release, search releases
3. Storage APIs: download/upload release artifact(s)

I think these groups roughly map to a natural division of services. For example, if a registry instance uses a replicated RDBMS, the "Maintainer APIs" would write to the primary store, "Consumer APIs" would read from replicas, and "Storage APIs" could use a general-purpose object store.

This would also allow for the "mirrored snapshots" of [user story](https://github.com/bytecodealliance/SIG-Registries/blob/main/docs/user-stories.md) F, by implementing just "Consumer" and/or "Storage" APIs.

**Comment by lann** [comment]
There is another cross-cutting versioning issue to consider: any maintainer-signed data (e.g. the "[release manifest](https://github.com/bytecodealliance/SIG-Registries/blob/proposal/publish-api/proposals/Publish-API.md#create-unpublished-release)" in the Publish API proposal) will be effectively immutable. We _may_ want to version this data schema separately from protocols to allow for the coexistence of older and newer signed data in a single protocol version.

**Reply by guybedford** [reply]
I'm not sure we'd want to permit metadata changes after publish time though? To dig into the extremely hypothetical example of a hash upgrade, a new message type like a "rehash" transaction signed by the registry itself could be used to relate the old hash to the new one in a way that is verifiable that both hashes hash the same content to add security without replacing the old data. The new publish API can be updated to only accept hashes of the new form, and eventually a major deprecation could drop the old hash format.

Are there other scenarios for maintainer-signed data updates that you had in mind here?
